typescript/rebalance_bot/swapdata.ts

```diff
@@ -0,0 +1,224 @@
+import { BotConfig, RebalanceQuote } from "../../config/types";
+import {
+  getSlippageLimitBps,
+  getMaxPriceImpactBps
+} from "../../config/constants";
+import { logger } from "../common/log";
+import { formatTokenAmount } from "../common/erc20";
+import { OdosClient } from "../odos/client";
+import { ContractManager } from "./contracts";
+
+export class SwapDataBuilder {
+  private odosClient: OdosClient;
+
+  constructor(
+    private readonly contracts: ContractManager,
+    private readonly config: BotConfig,
+  ) {
+    this.odosClient = new OdosClient(
+      "https://api.odos.xyz",
+      config.network.chainId,
+    );
+  }
+
+  async buildSwapData(
+    quote: RebalanceQuote,
+    trialRebalanceAmount: bigint,
+    userAddress: string,
+  ): Promise<string> {
+    if (!this.config.contracts.odosRouter) {
+      throw new Error("Odos router not configured for this network");
+    }
+
+    try {
+      if (quote.direction === 1) {
+        // Increase leverage: swap debt -> collateral (exact out)
+        return await this.buildIncreaseSwapData(
+          trialRebalanceAmount,
+          userAddress,
+        );
+      } else {
+        // Decrease leverage: swap collateral -> debt (exact out, including flash fee)
+        return await this.buildDecreaseSwapData(
+          trialRebalanceAmount,
+          userAddress,
+        );
+      }
+    } catch (error) {
+      logger.error("Failed to build swap data:", error);
+      throw error;
+    }
```

Comment:
Good addition! The price impact validation adds an important safety check. The configurable threshold via environment variable is a nice touch for flexibility.